trustfall_core/src/ir/ty.rs

```diff
@@ -0,0 +1,332 @@
+            }
+            if !ty_inside_list.nullable {
+                mask |= Modifiers::NON_NULLABLE_MASK << i;
+            }
+            base = &ty_inside_list.base;
+        }
+
+        let BaseType::Named(name) = base else {
+            unreachable!(
+                "should be impossible to get a non-named type after looping through all list types"
+            )
+        };
+        Type { base: name.to_string().into(), modifiers: Modifiers { mask } }
+    }
+}
+
+impl Display for Type {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // left
+        {
+            let mut current = Some(self.modifiers.clone());
+            while let Some(mods) = current {
+                if mods.is_list() {
+                    write!(f, "[")?;
+                }
+                current = mods.as_list();
+            }
+        }
+
+        write!(f, "{}", self.base)?;
+
+        let mut builder = String::new();
+
+        // right
+        {
+            let mut current = Some(self.modifiers.clone());
+            while let Some(mods) = current {
+                if !mods.nullable() {
+                    builder.push('!');
+                }
+                if mods.is_list() {
+                    builder.push(']');
+                }
+                current = mods.as_list();
+            }
+            write!(f, "{}", builder.chars().rev().collect::<String>())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Serialize for Type {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for Type {
+    fn deserialize<D>(deserializer: D) -> Result<Type, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct TypeDeserializer;
+
+        impl<'de> Visitor<'de> for TypeDeserializer {
+            type Value = Type;
+
+            fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
+                formatter.write_str("GraphQL type")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let ty = Type::new(s)
+                    .ok_or_else(|| serde::de::Error::custom("not a valid GraphQL type"))?;
+                Ok(ty)
+            }
+        }
+
+        deserializer.deserialize_str(TypeDeserializer)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::ir::ty::Type;
+
+    use super::Modifiers;
+
+    #[test]
+    fn max_allowed_nested_lists() {
+        let my_str = format!(
+            "{}String{}",
+            "[".repeat(Modifiers::MAX_LIST_DEPTH as usize),
+            "]".repeat(Modifiers::MAX_LIST_DEPTH as usize)
+        );
+        let constructed_type = Type::new(&my_str).unwrap().modifiers;
+        assert!(constructed_type.at_max_list_depth());
+    }
```

Comment:
Can you add a test like this but that also uses nullability?

Remember that a list contents' nullable bit is shifted one to the left from the last list bit. So while I think the implementation is probably correct, it's in principle possible that the nullability bit overflows even though the max list bit is fine.

Reply:
> Can you add a test like this but that also uses nullability?
Yes
> it's in principle possible that the nullability bit overflows even though the max list bit is fine.
I don't believe so. Here's the bit representation for that test `101010101010101010101010101010101010101010101010101010101010`
Here it is if there is an `!` at the very end too
`101010101010101010101010101010101010101010101010101010101011`

As you can see, only the very last bit on the very right changed from a zero to one, however I have added a mask representation test to ensure that doesn't happen.

Reply:
Ohh interesting, yeah you're right. If the innermost type is non-nullable, then we get the bitmask of:
`1101010101010101010101010101010101010101010101010101010101010`
(with no non-nulls)
`101010101010101010101010101010101010101010101010101010101010`
this does indeed add a new digit but we fit it fine.